crates/nu-protocol/src/config/completions.rs

```diff
@@ -0,0 +1,151 @@
+use super::{config_update_string_enum, prelude::*};
+use crate as nu_protocol;
+use crate::engine::Closure;
+
+#[derive(Clone, Copy, Debug, Default, IntoValue, PartialEq, Eq, Serialize, Deserialize)]
+pub enum CompletionAlgorithm {
+    #[default]
+    Prefix,
+    Fuzzy,
+}
+
+impl FromStr for CompletionAlgorithm {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_ascii_lowercase().as_str() {
+            "prefix" => Ok(Self::Prefix),
+            "fuzzy" => Ok(Self::Fuzzy),
+            _ => Err("'prefix' or 'fuzzy'"),
+        }
+    }
+}
+
+impl UpdateFromValue for CompletionAlgorithm {
+    fn update(&mut self, value: &Value, path: &mut ConfigPath, errors: &mut ConfigErrors) {
+        config_update_string_enum(self, value, path, errors)
+    }
+}
+
+#[derive(Clone, Copy, Debug, Default, IntoValue, PartialEq, Eq, Serialize, Deserialize)]
+pub enum CompletionSort {
+    #[default]
+    Smart,
+    Alphabetical,
+}
+
+impl FromStr for CompletionSort {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_ascii_lowercase().as_str() {
+            "smart" => Ok(Self::Smart),
+            "alphabetical" => Ok(Self::Alphabetical),
+            _ => Err("'smart' or 'alphabetical'"),
+        }
+    }
+}
+
+impl UpdateFromValue for CompletionSort {
+    fn update(&mut self, value: &Value, path: &mut ConfigPath, errors: &mut ConfigErrors) {
+        config_update_string_enum(self, value, path, errors)
+    }
+}
+
+#[derive(Clone, Debug, IntoValue, Serialize, Deserialize)]
+pub struct ExternalCompleterConfig {
+    pub enable: bool,
+    pub max_results: i64,
+    pub completer: Option<Closure>,
+}
+
+impl Default for ExternalCompleterConfig {
+    fn default() -> Self {
+        Self {
+            enable: true,
+            max_results: 100,
+            completer: None,
+        }
+    }
+}
+
+impl UpdateFromValue for ExternalCompleterConfig {
+    fn update<'a>(
+        &mut self,
+        value: &'a Value,
+        path: &mut ConfigPath<'a>,
+        errors: &mut ConfigErrors,
+    ) {
+        let Value::Record { val: record, .. } = value else {
+            errors.type_mismatch(path, Type::record(), value);
+            return;
+        };
+
+        for (col, val) in record.iter() {
+            let path = &mut path.push(col);
```

Comment:
This whole approach is certainly more modular but seeing this boilerplate all over the place also dilutes the important stuff.

Reply:
Yea... but the old code also had boilerplate everywhere and was actually unreadable :smile:
In the future, a derive macro would be nice so that all of this code could be removed.